### PR TITLE
#178 - Using exclusive storage access on metadata copy

### LINKED ISSUE #178: Avoid race conditions on adding artifact
When artifact is added to Maven repository at same location by multiple clients race conditions are possible. To avoid them it is possible to use exclusive access to certain part of `Storage`.

### PR DESCRIPTION
**Pull request by olegmoz**
Closes #178 
Using exclusive storage access on metadata copy

### CODE REVIEW

File: src/test/java/com/artipie/maven/http/UpdateMavenSliceTest.java [hunk 1]
```diff
@@ -189,4 +197,57 @@ void doesNotStartUpdateWhenSnapshotMetadataIsSent() {
...
+            } catch (final Exception ignored) {
+            }
+        }
+        MatcherAssert.assertThat(
+            "Some updates failed",
+            tasks.stream().anyMatch(CompletableFuture::isCompletedExceptionally),
+            new IsEqual<>(true)
+        );
+        MatcherAssert.assertThat(
+            "Storage has no locks",
```
Comment (g4s8) [inline src/test/java/com/artipie/maven/http/UpdateMavenSliceTest.java:247]: @olegmoz it's a test failure message, so it describes negative scenario, e.g. `"Storage has locks"`

**Review by g4s8 — CHANGES_REQUESTED** [review]
@olegmoz please check one minor comment

File: src/test/java/com/artipie/maven/http/UpdateMavenSliceTest.java (hunk 1 above)
Comment (olegmoz) [reply src/test/java/com/artipie/maven/http/UpdateMavenSliceTest.java:247]: @g4s8 thanks for pointing at that. After adding this check I noticed that some of exceptions are not about locks and point to a bug in ASTO, see https://github.com/artipie/asto/issues/260 

Going to fix it and then come back to this PR.

Comment (olegmoz) [reply src/test/java/com/artipie/maven/http/UpdateMavenSliceTest.java:284]: @g4s8 fixed the problem with locks and improved tests, please review

**Comment by olegmoz** [comment]
@olenagerasimova could you please review this issue?

**Review by g4s8 — APPROVED** [review]
@olegmoz thanks

**Comment by 0crat** [comment]
Job `gh:artipie/maven-adapter#179` is not assigned, can't get performer

**Pull request merged**